{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(GenomicRanges)\n",
    "library(ggplot2)\n",
    "library(TxDb.Hsapiens.UCSC.hg19.knownGene)\n",
    "library(ChIPseeker)\n",
    "library(org.Hs.eg.db)\n",
    "library(pheatmap)\n",
    "library(gplots)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Panel A - ucsc feature annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#eQTL data\n",
    "all <- readRDS(\"data/fig3A_all_GReQTLs_GRanges.rds\")\n",
    "female <- readRDS(\"data/fig3A_female_GReQTLs_GRanges.rds\")\n",
    "male <- readRDS(\"data/fig3A_male_GReQTLs_GRanges.rds\")\n",
    "\n",
    "# annotate peak\n",
    "ap.all<- annotatePeak(unique(all), TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene, annoDb = \"org.Hs.eg.db\")\n",
    "ap.female<- annotatePeak(unique(female), TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene, annoDb = \"org.Hs.eg.db\")\n",
    "ap.male<- annotatePeak(unique(male), TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene, annoDb = \"org.Hs.eg.db\")\n",
    "\n",
    "plot_all= ap.all@annoStat\n",
    "plot_all$Phenotype =\"all\"\n",
    "plot_female= ap.female@annoStat\n",
    "plot_female$Phenotype =\"female\"\n",
    "plot_male= ap.male@annoStat\n",
    "plot_male$Phenotype =\"male\"\n",
    "\n",
    "data_for_plot= rbind(plot_all, plot_female, plot_male)\n",
    "data_for_plot$Feature <- factor(data_for_plot$Feature, levels = c(\"1st Intron\",\n",
    "                                                                  \"Other Intron\",\n",
    "                                                                  \"1st Exon\",#\n",
    "                                                                  \"Other Exon\",\n",
    "                                                                  \"5' UTR\",#\n",
    "                                                                  \"3' UTR\",\n",
    "                                                                  \"Promoter (<=1kb)\",\n",
    "                                                                  \"Promoter (1-2kb)\",\n",
    "                                                                  \"Promoter (2-3kb)\",\n",
    "                                                                  \"Downstream (<=300)\",\n",
    "                                                                  \"Downstream (<=3kb)\",\n",
    "                                                                  \"Distal Intergenic\"))\n",
    "data_for_plot= rbind(data_for_plot, c(\"1st Exon\",\"0\", \"male noLD\")) #this is to add 1st exon to ledgend\n",
    "\n",
    "cat(\"2\")\n",
    "#pdf(\"Annotation_Fig3A.pdf\")\n",
    "ggplot(data_for_plot[ -(grep(\"noLD\",data_for_plot$Phenotype)),], aes(Phenotype, as.numeric(Frequency), fill = Feature)) +\n",
    "  geom_bar(position = \"fill\", stat = \"identity\") + scale_fill_brewer(name='Feature', palette = 'Paired') +\n",
    "  theme_minimal() + theme(legend.position=\"bottom\") + coord_flip()+ylab(\"Frequency\")+xlab(\"\")\n",
    "#dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Panel B - DeepSea feature annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds.data <- setNames(data.frame(matrix(ncol = 4, nrow = 0)), c(\"Feature\", \"Frequency\", \"Gender\", \"Group\"))\n",
    "ds.female <- read.delim(\"data/fig3B_female_deepSea.evalue\",head=T, sep=\",\")\n",
    "dim(ds.female)#1218  925\n",
    "ds.male <- read.delim(\"data/fig3B_male_deepSea.evalue\", head=T, sep=\",\")\n",
    "dim(ds.male)#[1] 2199  925\n",
    "fe_sig=ds.female[,-c(1:6)]\n",
    "fe_sig[fe_sig > 0.01] = 0\n",
    "fe_sig[fe_sig!=0] = 1\n",
    "fe_sig$sum=apply(fe_sig, 1,sum)\n",
    "fe_sig.percent = (dim(fe_sig[fe_sig$sum>0,])[1]*100)/dim(ds.female)[1]\n",
    "ds.data[1,] = c(\"DeepSea features\", fe_sig.percent, \"females\", \"DeepSea features\")\n",
    "ds.data[2,] = c(\"no DeepSea features\", 100-fe_sig.percent, \"females\", \"DeepSea features\")\n",
    "m_sig=ds.male[,-c(1:6)]\n",
    "m_sig[m_sig > 0.01]=0\n",
    "m_sig[m_sig!=0]=1\n",
    "m_sig$sum=apply(m_sig, 1,sum)\n",
    "dim(m_sig[m_sig$sum>0,])#851 = 38.69941%\n",
    "m_sig.percent = (dim(m_sig[m_sig$sum>0,])[1]*100)/dim(ds.male)[1]\n",
    "ds.data[3,] = c(\"DeepSea features\", m_sig.percent, \"males\", \"DeepSea features\")\n",
    "ds.data[4,] = c(\"no DeepSea features\", 100-m_sig.percent, \"males\", \"DeepSea features\")\n",
    "# - GRE annotation\n",
    "gre.female <- readRDS(\"data/fig3B_female_GReQTLs_GRE.rds\")\n",
    "fe_gre.percent = (length(gre.female@seqnames)*100)/length(female@seqnames)\n",
    "ds.data[5,] = c(\"GRE\", fe_gre.percent, \"females\", \"GRE\")\n",
    "ds.data[6,] = c(\"no GRE\", 100-fe_gre.percent, \"females\", \"GRE\")\n",
    "\n",
    "gre.male <- readRDS(\"data/fig3B_male_GReQTLs_GRE.rds\")\n",
    "m_gre.percent = (length(gre.male@seqnames)*100)/length(male@seqnames)\n",
    "ds.data[7,] = c(\"GRE\", m_gre.percent, \"males\", \"GRE\")\n",
    "ds.data[8,] = c(\"no GRE\", 100-m_gre.percent, \"males\", \"GRE\")\n",
    "# - ARE and ER annotation\n",
    "are.female <- readRDS(\"data/fig3B_female_GReQTLs_ARE.rds\")\n",
    "fe_are.percent = (length(are.female@seqnames)*100)/length(female@seqnames)\n",
    "ds.data[9,] = c(\"ARE\", fe_are.percent, \"females\", \"ARE\")\n",
    "ds.data[10,] = c(\"no ARE\", 100-fe_are.percent, \"females\", \"ARE\")\n",
    "ere.female <- readRDS(\"data/fig3B_female_GReQTLs_ERE.rds\")\n",
    "fe_ere.percent = (length(ere.female@seqnames)*100)/length(female@seqnames)\n",
    "ds.data[11,] = c(\"ERE\", fe_ere.percent, \"females\", \"ERE\")\n",
    "ds.data[12,] = c(\"no ERE\", 100-fe_ere.percent, \"females\", \"ERE\")\n",
    "\n",
    "are.male <- readRDS(\"data/fig3B_male_GReQTLs_ARE.rds\")\n",
    "fe_are.percent = (length(are.male@seqnames)*100)/length(male@seqnames)\n",
    "ds.data[13,] = c(\"ARE\", fe_are.percent, \"males\", \"ARE\")\n",
    "ds.data[14,] = c(\"no ARE\", 100-fe_are.percent, \"males\" ,\"ARE\")\n",
    "ere.male <- readRDS(\"data/fig3B_male_GReQTLs_ERE.rds\")\n",
    "fe_ere.percent = (length(ere.male@seqnames)*100)/length(male@seqnames)\n",
    "ds.data[15,] = c(\"ERE\", fe_ere.percent, \"males\", \"ERE\")\n",
    "ds.data[16,] = c(\"no ERE\", 100-fe_ere.percent, \"males\", \"ERE\")\n",
    "\n",
    "ds.data$Feature <- factor(ds.data$Feature, levels = c(\"no DeepSea features\",\"DeepSea features\",\"no GRE\",\"GRE\",\"no ARE\", \"ARE\",\"no ERE\",\"ERE\")  )          \n",
    "#pdf(\"Annotation_Fig3B.pdf\")\n",
    "ggplot(ds.data, aes(Gender, as.numeric(Frequency), fill = Feature)) +\n",
    "  geom_bar(position = \"fill\", stat = \"identity\") + scale_fill_brewer(name='Feature', palette = 'Paired') +\n",
    "  theme_minimal() + theme(legend.position=\"bottom\") + coord_flip()+ylab(\"Frequency\")+xlab(\"\")+facet_wrap(~Group)\n",
    "#dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Panel C - ChromHMM annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fe.p <-read.delim(\"data/fig3C_female_chromHMM_pMatrix.txt\", head=T, check.names = F)\n",
    "fe.or <- read.delim(\"data/fig3C_female_chromHMM_orMatrix.txt\", head=T, check.names = F)\n",
    "m.p <- read.delim(\"data/fig3C_male_chromHMM_pMatrix.txt\", head=T, check.names = F)\n",
    "m.or <- read.delim(\"data/fig3C_male_chromHMM_orMatrix.txt\", head=T, check.names = F)\n",
    "codes <- read.delim(\"data/fig3C_chromHMM_codes.txt\", head=T)\n",
    "\n",
    "# pheatmap\n",
    "colnames(fe.or)[-c(1:2)]= paste0(\"females_\",sapply(colnames(fe.or)[-c(1:2)], function(x) unlist(strsplit(x,\"_\"))[2]))\n",
    "colnames(m.or)[-c(1:2)]= paste0(\"males_\",sapply(colnames(m.or)[-c(1:2)], function(x) unlist(strsplit(x,\"_\"))[2]))\n",
    "colnames(fe.p)[-c(1:2)]= paste0(\"females_\",sapply(colnames(fe.p)[-c(1:2)], function(x) unlist(strsplit(x,\"_\"))[2]))\n",
    "colnames(m.p)[-c(1:2)]= paste0(\"males\",sapply(colnames(m.p)[-c(1:2)], function(x) unlist(strsplit(x,\"_\"))[2]))\n",
    "\n",
    "both_or=merge(fe.or, m.or, by=c(\"EID\", \"Standardized.Epigenome.name\"))\n",
    "both_p=merge(fe.p, m.p, by=c(\"EID\", \"Standardized.Epigenome.name\"))\n",
    "both_or[is.na(both_or)]<-0\n",
    "both_or[both_or==\"Inf\"]<-0\n",
    "both_or[both_or==\"-Inf\"]<-0\n",
    "\n",
    "#select blood cells only\n",
    "codes.sub=codes[codes$GROUP %in% c( \"Blood & T-cell\"),]$EID\n",
    "both_or_sub=both_or[ both_or$EID %in% codes.sub,]\n",
    "both_p_sub=both_p[ both_p$EID %in% codes.sub,]\n",
    "both_or_sort=both_or_sub[ ,sort(colnames(both_or_sub))]\n",
    "both_p_sort=both_p_sub[ ,sort(colnames(both_p_sub))]\n",
    "\n",
    "# heatmap\n",
    "breaks=c(seq(0,8,by=0.02))\n",
    "color.palette  <- colorRampPalette(c(\"white\", \"#2C7BB6\"))(length(breaks) - 1)\n",
    "plot_or=both_or_sort[,-c(32,1)]\n",
    "plot_p=both_p_sort[,-c(32,1)]\n",
    "rownames(plot_or)=both_or_sort$Standardized.Epigenome.name\n",
    "rownames(plot_p)=both_p_sort$Standardized.Epigenome.name\n",
    "\n",
    "#plot only sig. p-values\n",
    "plot_p[plot_p>0.05] = \"\"\n",
    "#pdf(\"Annotation_Fig3C.pdf\")\n",
    "heatmap.2((as.matrix(plot_or)), cellnote=(as.matrix(plot_p)),\n",
    "          notecol=\"black\",col=color.palette,scale=\"none\",key=TRUE, keysize=1, Rowv = F,Colv = F,dendrogram = \"none\",\n",
    "          density.info=\"none\", trace=\"none\", cexRow=0.5,cexCol=0.5, notecex =0.2)\n",
    "#dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
